packages/boot/src/boot.mixin.ts

```diff
@@ -0,0 +1,129 @@
+ * - Adds an optional `bootOptions` property to the Class that can be used to
+ *    store the Booter conventions.
+ * - Adds the `BootComponent` to the Class (which binds the Bootstrapper and default Booters)
+ * - Provides the `boot()` convenience method to call Bootstrapper.boot()
+ * - Provides the `booter()` convenience method to bind a Booter(s) to the Application
+ * - Override `component()` to call `mountComponentBooters`
+ * - Adds `mountComponentBooters` which binds Booters to the application from `component.booters[]`
+ */
+// tslint:disable-next-line:no-any
+export function BootMixin<T extends Constructor<any>>(superClass: T) {
+  return class extends superClass implements Bootable {
+    projectRoot: string;
+    bootOptions?: BootOptions;
+
+    // tslint:disable-next-line:no-any
+    constructor(...args: any[]) {
+      super(...args);
+      this.component(BootComponent);
+    }
+
+    /**
+     * Convenience method to call bootstrapper.boot()
+     */
+    async boot(): Promise<void> {
+      // Project Root and Boot Options need to be bound so we can resolve an
+      // instance of Bootstrapper.
+      this.bind(BootBindings.PROJECT_ROOT).to(this.projectRoot);
+      this.bind(BootBindings.BOOT_OPTIONS).to(this.bootOptions);
+
+      // Get a instance of the BootStrapper
+      const bootstrapper: Bootstrapper = await this.get(
+        BootBindings.BOOTSTRAPPER_KEY,
+      );
+
+      await bootstrapper.boot();
+    }
+
+    /**
+     * Given a N number of Booter Classes, this method binds them using the
+     * prefix and tag expected by the Bootstrapper.
+     *
+     * @param booterCls Booter classes to bind to the Application
+     *
+     * ```ts
+     * app.booter(MyBooter, MyOtherBooter)
+     * ```
+     */
+    booter(...booterCls: Constructor<Booter>[]): Binding[] {
+      // tslint:disable-next-line:no-any
+      return booterCls.map(cls => _bindBooter(<Context>(<any>this), cls));
+    }
+
+    /**
+     * Override to ensure any Booter's on a Component are also mounted.
+     *
+     * @param component The component to add.
+     *
+     * ```ts
+     *
+     * export class ProductComponent {
+     *   booters = [ControllerBooter, RepositoryBooter];
+     *   providers = {
+     *     [AUTHENTICATION_STRATEGY]: AuthStrategy,
+     *     [AUTHORIZATION_ROLE]: Role,
+     *   };
+     * };
+     *
+     * app.component(ProductComponent);
+     * ```
+     */
+    public component(component: Constructor<{}>) {
+      super.component(component);
+      this.mountComponentBooters(component);
+    }
+
+    /**
+     * Get an instance of a component and mount all it's
+     * booters. This function is intended to be used internally
+     * by component()
+     *
+     * @param component The component to mount booters of
+     */
+    mountComponentBooters(component: Constructor<{}>) {
+      const componentKey = `components.${component.name}`;
+      const compInstance = this.getSync(componentKey);
```

Comment:
This pattern when we have `mountComponent{something}` calling `this.getSync` to obtain the component instance is starting to repeat in too many places. IMO, we should refactor `app.component` and introduce `app.mountComponent` that can be used to access things provided by component instances.

```ts
// in application
  public component(componentCtor: Constructor<Component>, name?: string) {
    name = name || componentCtor.name;
    const componentKey = `components.${name}`;
    this.bind(componentKey)
      .toClass(componentCtor)
      .inScope(BindingScope.SINGLETON)
      .tag('component');
    // Assuming components can be synchronously instantiated
    const instance = this.getSync(componentKey);
    this._mountComponent(instance);
  }

  protected _mountComponent() {
    mountComponent(this, instance);
  }

  // in extensions/component mixins:
  // - do not change public component() method
  // - change mountComponent instead
  protected _mountComponent(component: Component & {booters: /*add typedef*/}) {
    super._mountComponent(component);
    if (component.booters) {
      this.booter(...compInstance.booters);
    }
  }
```

Thoughts? Feel free to leave this improvement out of this PR.

Reply:
I think this is a great idea! I've created https://github.com/strongloop/loopback-next/issues/1017 to track this and keep it out of this PR. Feel free to edit that issue if I missed anything.